hivemind/p2p/p2p_daemon_bindings/control.py

```diff
@@ -90,8 +105,130 @@ async def listen(self) -> AsyncIterator["ControlClient"]:
         else:
             raise ValueError(f"Protocol not supported: {protocols.protocol_with_code(proto_code)}")
 
-        async with server:
-            yield self
+        try:
+            async with server:
+                yield self
+        finally:
+            if self._read_task is not None:
+                self._read_task.cancel()
+            if self._write_task is not None:
+                self._write_task.cancel()
+
+    async def _read_from_persistent_conn(self, reader: asyncio.StreamReader):
+        with closing(reader):
+            while True:
+                resp = p2pd_pb.PCResponse()
+                await read_pbmsg_safe(reader, resp)
+
+                call_id = uuid.UUID(bytes=resp.callId)
+
+                if resp.HasField("callUnaryResponse"):
+                    if call_id in self.pending_calls and resp.callUnaryResponse.HasField("response"):
+                        self.pending_calls[call_id].set_result(resp.callUnaryResponse.response)
+                    elif call_id in self.pending_calls and resp.callUnaryResponse.HasField("error"):
+                        remote_exc = P2PHandlerError(resp.callUnaryResponse.error.decode())
```

Comment:
```suggestion
                        remote_exc = P2PHandlerError(resp.callUnaryResponse.error.decode(errors='ignore'))
```

Otherwise, a remote party may crash the daemon by sending the error with invalid UTF-8 sequence.